File: clients/desktop/Tests/TrustStoreTests.cs
using System;
using Aura.Desktop.Services;
using Xunit;

namespace Aura.Desktop.Tests;

/// <summary>
/// Trust bookkeeping behind the trust sheet. The rules come from the design
/// handoff's §7 ("three states only: verified / TOFU · not compared / key
/// changed today") and its state table, where verifiedKeys is described as the
/// thing that "drives every trust badge".
///
/// The property worth defending here is that trust attaches to a <em>key</em>,
/// not to a name — otherwise a takeover inherits the trust of whoever held the
/// name before.
/// </summary>
public class TrustStoreTests
{
    private sealed class FakeClock
    {
        public DateTimeOffset Now = new(2026, 1, 1, 12, 0, 0, TimeSpan.Zero);
        public DateTimeOffset Get() => Now;
        public void AdvanceDays(int d) => Now = Now.AddDays(d);
    }

    private const string Alice = "6ba7b810-9dad-11d1-80b4-00c04fd430c8";
    private const string Bob = "6ba7b811-9dad-11d1-80b4-00c04fd430c8";

    private static byte[] Key(byte fill) => System.Linq.Enumerable.ToArray(
        System.Linq.Enumerable.Repeat(fill, 32));

    private static (TrustStore store, AppSettings settings, FakeClock clock, Func<int> saves) New()
    {
        var settings = new AppSettings();
        var clock = new FakeClock();
        var saveCount = 0;
        var store = new TrustStore(settings, () => saveCount++, clock.Get);
        return (store, settings, clock, () => saveCount);
    }

    [Fact]
    public void UnknownKey_StartsAsTofu()
    {
        var (store, _, _, _) = New();
        Assert.Equal(TrustState.Tofu, store.Evaluate(Alice, Key(0x11)));
    }

    [Fact]
    public void ObservedThenVerified_BecomesVerified()
    {
        var (store, _, _, _) = New();
        var key = Key(0x11);

        store.Observe(Alice, key);
        Assert.Equal(TrustState.Tofu, store.Evaluate(Alice, key));

        store.MarkVerified(Alice, key);
        Assert.Equal(TrustState.Verified, store.Evaluate(Alice, key));
    }

    [Fact]
    public void NewKeyForKnownUser_ReportsKeyChanged()
    {
        var (store, _, _, _) = New();

        store.Observe(Alice, Key(0x11));
        store.Observe(Alice, Key(0x22));   // reinstall, or a takeover

        Assert.Equal(TrustState.KeyChanged, store.Evaluate(Alice, Key(0x22)));
    }

    [Fact]
    public void VerifyingOneKeyDoesNotVerifyAnother()
    {
        // The whole point of keying on the key: trust must not transfer to a
        // different key just because the same person presents it.
        var (store, _, _, _) = New();

        store.Observe(Alice, Key(0x11));
        store.MarkVerified(Alice, Key(0x11));

        Assert.Equal(TrustState.Verified, store.Evaluate(Alice, Key(0x11)));
        Assert.NotEqual(TrustState.Verified, store.Evaluate(Alice, Key(0x22)));
    }

    [Fact]
    public void TrustDoesNotTransferBetweenUsersSharingAName()
    {
        // Bob has never been seen; presenting Alice's verified key does make
        // that key verified (trust is per key), but Bob's own key must not be.
        var (store, _, _, _) = New();

        store.Observe(Alice, Key(0x11));
        store.MarkVerified(Alice, Key(0x11));

        Assert.Equal(TrustState.Tofu, store.Evaluate(Bob, Key(0x99)));
    }

    [Fact]
    public void VerifyingTheNewKeyClearsKeyChanged()
    {
        var (store, _, _, _) = New();

        store.Observe(Alice, Key(0x11));
        store.Observe(Alice, Key(0x22));
        Assert.Equal(TrustState.KeyChanged, store.Evaluate(Alice, Key(0x22)));

        store.MarkVerified(Alice, Key(0x22));

        Assert.Equal(TrustState.Verified, store.Evaluate(Alice, Key(0x22)));
        Assert.Null(store.KeyChangedAt(Alice));
    }

    [Fact]
    public void ReturningToTheBaselineKeyClearsTheChangedMarker()
    {
        // Someone switches devices and switches back. The warning should not
        // outlive the condition that raised it.
        var (store, _, _, _) = New();

        store.Observe(Alice, Key(0x11));
        store.Observe(Alice, Key(0x22));
        Assert.NotNull(store.KeyChangedAt(Alice));

        store.Observe(Alice, Key(0x11));

        Assert.Null(store.KeyChangedAt(Alice));
        Assert.Equal(TrustState.Tofu, store.Evaluate(Alice, Key(0x11)));
    }

    [Fact]
    public void KeyChangedAt_StampsFirstSightingOnly()
    {
        var (store, _, clock, _) = New();

        store.Observe(Alice, Key(0x11));
        store.Observe(Alice, Key(0x22));
        var first = store.KeyChangedAt(Alice);

        clock.AdvanceDays(3);
        store.Observe(Alice, Key(0x22));   // still the same disagreement

        Assert.Equal(first, store.KeyChangedAt(Alice));
    }

    [Fact]
    public void BlockOutranksVerified()
    {
        var (store, _, _, _) = New();
        var key = Key(0x11);

        store.MarkVerified(Alice, key);
        store.Block(key);

        Assert.Equal(TrustState.Blocked, store.Evaluate(Alice, key));
    }

    [Fact]
    public void BlockAppliesToTheKeyNotThePerson()
    {
        var (store, _, _, _) = New();

        store.Observe(Alice, Key(0x11));
        store.MarkVerified(Alice, Key(0x11));
        store.Block(Key(0x22));

        // The impostor key is blocked; the key Alice actually verified is not.
        Assert.Equal(TrustState.Blocked, store.Evaluate(Alice, Key(0x22)));
        Assert.Equal(TrustState.Verified, store.Evaluate(Alice, Key(0x11)));
    }

    [Fact]
    public void VerifyingAfterBlockClearsTheBlock()
    {
        var (store, _, _, _) = New();
        var key = Key(0x11);

        store.Block(key);
        store.MarkVerified(Alice, key);

        Assert.Equal(TrustState.Verified, store.Evaluate(Alice, key));
    }

    [Fact]
    public void UnblockRestoresThePreviousState()
    {
        var (store, _, _, _) = New();
        var key = Key(0x11);

        store.Observe(Alice, key);
        store.Block(key);
        store.Unblock(key);

        Assert.Equal(TrustState.Tofu, store.Evaluate(Alice, key));
    }

    [Fact]
    public void DecisionsArePersistedThroughSettings()
    {
        // A verification is an out-of-band ritual; losing it to a crash means
        // asking the user to do it again, so every mutation writes through.
        var (store, settings, clock, saves) = New();
        var key = Key(0x11);

        store.MarkVerified(Alice, key);

        var id = TrustStore.KeyId(key);
        Assert.True(settings.VerifiedKeys.ContainsKey(id));
        Assert.Equal(clock.Get(), settings.VerifiedKeys[id]);
        Assert.Equal(id, settings.KnownUserKeys[Alice]);
        Assert.True(saves() > 0);
    }

    [Fact]
    public void ReloadedSettingsPreserveVerification()
    {
        var (store, settings, _, _) = New();
        var key = Key(0x11);
        store.MarkVerified(Alice, key);

        // Same dictionaries, fresh store — stands in for a restart.
        var reloaded = new TrustStore(settings, () => { });

        Assert.Equal(TrustState.Verified, reloaded.Evaluate(Alice, key));
    }

    [Fact]
    public void KeyIdIsLowercaseHex()
    {
        Assert.Equal(new string('1', 64), TrustStore.KeyId(Key(0x11)));
        Assert.Equal(new string('a', 64), TrustStore.KeyId(Key(0xAA)));
    }

    [Fact]
    public void EmptyInputsAreIgnoredRatherThanStored()
    {
        var (store, settings, _, _) = New();

        store.Observe(Alice, Array.Empty<byte>());
        store.Observe("", Key(0x11));
        store.MarkVerified(Alice, Array.Empty<byte>());
        store.Block(Array.Empty<byte>());

        Assert.Empty(settings.KnownUserKeys);
        Assert.Empty(settings.VerifiedKeys);
        Assert.Empty(settings.BlockedKeys);
        Assert.Equal(TrustState.Tofu, store.Evaluate(Alice, Array.Empty<byte>()));
    }
}

File: clients/macos/AuraTests/TrustStoreTests.swift
import XCTest
@testable import Aura

/// Trust bookkeeping behind the trust sheet. The rules come from the design
/// handoff's §7 ("three states only: verified / TOFU · not compared / key
/// changed today") and its state table, where verifiedKeys is described as the
/// thing that "drives every trust badge".
///
/// The property worth defending here is that trust attaches to a *key*, not to
/// a name — otherwise a takeover inherits the trust of whoever held the name
/// before.
final class TrustStoreTests: XCTestCase {

    private final class FakeClock {
        /// 2026-01-01T12:00:00Z
        var now = Date(timeIntervalSince1970: 1_767_268_800)
        func advanceDays(_ days: Int) { now = now.addingTimeInterval(TimeInterval(days) * 86_400) }
    }

    /// Stands in for UserDefaults so the tests never touch the real user's
    /// preferences, and counts writes so write-through can be asserted.
    private final class InMemoryDefaults: TrustStoreDefaults {
        private(set) var storage: [String: Data] = [:]
        private(set) var writes = 0

        func data(forKey defaultName: String) -> Data? { storage[defaultName] }

        func set(_ value: Any?, forKey defaultName: String) {
            writes += 1
            storage[defaultName] = value as? Data
        }
    }

    private let alice = "6ba7b810-9dad-11d1-80b4-00c04fd430c8"
    private let bob = "6ba7b811-9dad-11d1-80b4-00c04fd430c8"

    private func key(_ fill: UInt8) -> Data { Data(repeating: fill, count: 32) }

    private func newStore() -> (store: TrustStore, defaults: InMemoryDefaults, clock: FakeClock) {
        let defaults = InMemoryDefaults()
        let clock = FakeClock()
        let store = TrustStore(defaults: defaults, clock: { clock.now })
        return (store, defaults, clock)
    }

    func testUnknownKeyStartsAsTofu() {
        let (store, _, _) = newStore()
        XCTAssertEqual(store.evaluate(userUuid: alice, publicKey: key(0x11)), .tofu)
    }

    func testObservedThenVerifiedBecomesVerified() {
        let (store, _, _) = newStore()
        let key = key(0x11)

        store.observe(userUuid: alice, publicKey: key)
        XCTAssertEqual(store.evaluate(userUuid: alice, publicKey: key), .tofu)

        store.markVerified(userUuid: alice, publicKey: key)
        XCTAssertEqual(store.evaluate(userUuid: alice, publicKey: key), .verified)
    }

    func testNewKeyForKnownUserReportsKeyChanged() {
        let (store, _, _) = newStore()

        store.observe(userUuid: alice, publicKey: key(0x11))
        store.observe(userUuid: alice, publicKey: key(0x22))   // reinstall, or a takeover

        XCTAssertEqual(store.evaluate(userUuid: alice, publicKey: key(0x22)), .keyChanged)
    }

    func testVerifyingOneKeyDoesNotVerifyAnother() {
        // The whole point of keying on the key: trust must not transfer to a
        // different key just because the same person presents it.
        let (store, _, _) = newStore()

        store.observe(userUuid: alice, publicKey: key(0x11))
        store.markVerified(userUuid: alice, publicKey: key(0x11))

        XCTAssertEqual(store.evaluate(userUuid: alice, publicKey: key(0x11)), .verified)
        XCTAssertNotEqual(store.evaluate(userUuid: alice, publicKey: key(0x22)), .verified)
    }

    func testTrustDoesNotTransferBetweenUsersSharingAName() {
        // Bob has never been seen; presenting Alice's verified key does make
        // that key verified (trust is per key), but Bob's own key must not be.
        let (store, _, _) = newStore()

        store.observe(userUuid: alice, publicKey: key(0x11))
        store.markVerified(userUuid: alice, publicKey: key(0x11))

        XCTAssertEqual(store.evaluate(userUuid: bob, publicKey: key(0x99)), .tofu)
    }

    func testVerifyingTheNewKeyClearsKeyChanged() {
        let (store, _, _) = newStore()

        store.observe(userUuid: alice, publicKey: key(0x11))
        store.observe(userUuid: alice, publicKey: key(0x22))
        XCTAssertEqual(store.evaluate(userUuid: alice, publicKey: key(0x22)), .keyChanged)

        store.markVerified(userUuid: alice, publicKey: key(0x22))

        XCTAssertEqual(store.evaluate(userUuid: alice, publicKey: key(0x22)), .verified)
        XCTAssertNil(store.keyChangedAt(userUuid: alice))
    }

    func testReturningToTheBaselineKeyClearsTheChangedMarker() {
        // Someone switches devices and switches back. The warning should not
        // outlive the condition that raised it.
        let (store, _, _) = newStore()

        store.observe(userUuid: alice, publicKey: key(0x11))
        store.observe(userUuid: alice, publicKey: key(0x22))
        XCTAssertNotNil(store.keyChangedAt(userUuid: alice))

        store.observe(userUuid: alice, publicKey: key(0x11))

        XCTAssertNil(store.keyChangedAt(userUuid: alice))
        XCTAssertEqual(store.evaluate(userUuid: alice, publicKey: key(0x11)), .tofu)
    }

    func testKeyChangedAtStampsFirstSightingOnly() {
        let (store, _, clock) = newStore()

        store.observe(userUuid: alice, publicKey: key(0x11))
        store.observe(userUuid: alice, publicKey: key(0x22))
        let first = store.keyChangedAt(userUuid: alice)

        clock.advanceDays(3)
        store.observe(userUuid: alice, publicKey: key(0x22))   // still the same disagreement

        XCTAssertEqual(first, store.keyChangedAt(userUuid: alice))
    }

    func testBlockOutranksVerified() {
        let (store, _, _) = newStore()
        let key = key(0x11)

        store.markVerified(userUuid: alice, publicKey: key)
        store.block(publicKey: key)

        XCTAssertEqual(store.evaluate(userUuid: alice, publicKey: key), .blocked)
    }

    func testBlockAppliesToTheKeyNotThePerson() {
        let (store, _, _) = newStore()

        store.observe(userUuid: alice, publicKey: key(0x11))
        store.markVerified(userUuid: alice, publicKey: key(0x11))
        store.block(publicKey: key(0x22))

        // The impostor key is blocked; the key Alice actually verified is not.
        XCTAssertEqual(store.evaluate(userUuid: alice, publicKey: key(0x22)), .blocked)
        XCTAssertEqual(store.evaluate(userUuid: alice, publicKey: key(0x11)), .verified)
    }

    func testVerifyingAfterBlockClearsTheBlock() {
        let (store, _, _) = newStore()
        let key = key(0x11)

        store.block(publicKey: key)
        store.markVerified(userUuid: alice, publicKey: key)

        XCTAssertEqual(store.evaluate(userUuid: alice, publicKey: key), .verified)
    }

    func testUnblockRestoresThePreviousState() {
        let (store, _, _) = newStore()
        let key = key(0x11)

        store.observe(userUuid: alice, publicKey: key)
        store.block(publicKey: key)
        store.unblock(publicKey: key)

        XCTAssertEqual(store.evaluate(userUuid: alice, publicKey: key), .tofu)
    }

    func testDecisionsArePersistedThroughSettings() {
        // A verification is an out-of-band ritual; losing it to a crash means
        // asking the user to do it again, so every mutation writes through.
        let (store, defaults, clock) = newStore()
        let key = key(0x11)

        store.markVerified(userUuid: alice, publicKey: key)

        let id = TrustStore.keyId(key)
        XCTAssertNotNil(store.verifiedKeys[id])
        XCTAssertEqual(store.verifiedKeys[id], clock.now)
        XCTAssertEqual(store.knownUserKeys[alice], id)
        XCTAssertGreaterThan(defaults.writes, 0)
    }

    func testReloadedSettingsPreserveVerification() {
        let (store, defaults, _) = newStore()
        let key = key(0x11)
        store.markVerified(userUuid: alice, publicKey: key)

        // Same backing store, fresh store object — stands in for a restart.
        let reloaded = TrustStore(defaults: defaults)

        XCTAssertEqual(reloaded.evaluate(userUuid: alice, publicKey: key), .verified)
    }

    func testKeyIdIsLowercaseHex() {
        XCTAssertEqual(TrustStore.keyId(key(0x11)), String(repeating: "1", count: 64))
        XCTAssertEqual(TrustStore.keyId(key(0xAA)), String(repeating: "a", count: 64))
    }

    func testEmptyInputsAreIgnoredRatherThanStored() {
        let (store, _, _) = newStore()

        store.observe(userUuid: alice, publicKey: Data())
        store.observe(userUuid: "", publicKey: key(0x11))
        store.markVerified(userUuid: alice, publicKey: Data())
        store.block(publicKey: Data())

        XCTAssertTrue(store.knownUserKeys.isEmpty)
        XCTAssertTrue(store.verifiedKeys.isEmpty)
        XCTAssertTrue(store.blockedKeys.isEmpty)
        XCTAssertEqual(store.evaluate(userUuid: alice, publicKey: Data()), .tofu)
    }
}
